File: Python/algv3.py
#!/usr/bin/python3

import sys
import re
from datetime import datetime

def parseARGV(args):
  if args is None:
    return None
  arg=args[1]
  v=[]
  m=re.match(r'^-?\d+$',arg)
  if m is not None:
    n=len(args)
    for i in range(1,n):
      v.append(int(args[i]))
    return v
  infile=sys.stdin
  m=re.match(r'^-?-?[Ss][Tt][Dd](?:[Ii][Nn])?$',arg)
  cap=0
  if m is None:
    infile=open(arg,'r',encoding="utf-8")
  if len(sys.argv)>2:
    arg=args[2]
    m=re.match(r'^[1-9]\d*$',arg)
    if m is not None:
      cap=int(arg)
      v=[0]*cap
  if cap==0:
    for line in infile:
      v.append(int(line))
    return v
  x=0
  for line in infile:
    v[x]=int(line)
    x=x+1
    if x==cap:
      break
  for line in infile:
    v.append(int(line))
    x=x+1
  if x<cap:
    del v[x:]
  return v

#https://www.youtube.com/watch?v=Qgevy75co8c
def somaMaximaV3(v):
  if v is None:
    return None
  if len(v)==0:
    return None
  return somaMaximaV3Recursivo(v,0,len(v)-1)

def somaMaximaV3Recursivo(v,esquerda,direita):
  #se so tem uma posicao o resultado eh o valor desta posicao
  if esquerda==direita:
    return v[esquerda]

  #caso com mais de uma posicao
  meio=(esquerda+direita)//2

  #calcula soma maxima a esquerda do ponto mediano
  somaMaximaEsquerda=somaMaximaV3Recursivo(v,esquerda,meio)

  #calcula soma maxima a direita do ponto mediano
  somaMaximaDireita=somaMaximaV3Recursivo(v,meio+1,direita)

  #inicia soma parcial esquerda
  s=v[meio]

  #inicial resultado parcial esquerdo
  rEsquerda=v[meio]

  #para todas as posicoes a esquerda
  for i in range(meio-1,esquerda-1,-1):
    s=s+v[i]
    if s>rEsquerda:
      rEsquerda=s

  #inicia soma parcial direita
  s=v[meio+1]

  #inicial resultado parcial direito
  rDireita=v[meio+1]

  #para todas as posicoes a direita
  for i in range(meio+2,direita+1):
    s=s+v[i]
    if s>rDireita:
      rDireita=s

  r=max(somaMaximaEsquerda,rEsquerda+rDireita,somaMaximaDireita)
  return r

if len(sys.argv)<2:
  print(f'{sys.argv[0]} : vetor\n')
  sys.exit(1)

preparse=datetime.now()
v=parseARGV(sys.argv)
if v is None:
  print(f'{sys.argv[0]} : ERRO no parse de ({sys.argv[1]})\n')
  sys.exit(2)

precalc=datetime.now()
somaMaxima=somaMaximaV3(v)
poscalc=datetime.now()

aloctime=precalc-preparse
algtime=poscalc-precalc

print(f'\n{sys.argv[0]} : Tempo Alocacao memoria vetor ({aloctime.total_seconds()})')
print(f'\n{sys.argv[0]} : Tempo Calculo da soma maxima ({algtime.total_seconds()})')
print(f'\n{sys.argv[0]} : A soma maxima do array com ({len(v)}) posicoes eh ({somaMaxima})')

File: Perl/algv3.pl
#!/usr/bin/perl -w

use strict;
use Time::HiRes qw(time);

if(!@ARGV)
{
  die "\n$0 : vetor\n";
}

my $preparse=time();
if(&parseARGV($ARGV[0])<0)
{
  die "\n$0 : ERRO no parse de ($ARGV[0])\n";
}

my $precalc=time();
my $somaMaxima=&somaMaximaV3(\@ARGV);
if(!defined $somaMaxima)
{
  die "\n$0 : ERRO ao obter a soma maxima\n";
}
my $poscalc=time();
print "\n$0 : Tempo Alocacao memoria vetor (".($precalc-$preparse).")\n";
print "\n$0 : Tempo Calculo da soma maxima (".($poscalc-$precalc).")\n";
print "\n$0 : A soma maxima do array com (".scalar(@ARGV).") posicoes eh ($somaMaxima)\n";

sub somaMaximaV3
{
  my $v=shift;
  return &somaMaximaV3Recursivo(0,scalar(@$v)-1,$v);
}

sub somaMaximaV3Recursivo
{
  my $esquerda=shift;
  my $direita=shift;
  my $v=shift;

  my $n=scalar(@$v);
  if(!$n)
  {
    return undef;
  }

  #se so tem uma posicao o resultado eh o valor desta posicao
  if($esquerda==$direita)
  {
    return $v->[$esquerda];
  }

  #caso com mais de uma posicao
  my $meio=int(($esquerda+$direita)/2);

  #calcula soma maxima a esquerda do ponto mediano
  my $somaMaximaEsquerda=&somaMaximaV3Recursivo($esquerda,$meio,$v);

  #calcula soma maxima a direita do ponto mediano
  my $somaMaximaDireita=&somaMaximaV3Recursivo($meio+1,$direita,$v);

  #inicia soma parcial esquerda
  my $s=$v->[$meio];

  #inicial resultado parcial esquerdo
  my $rEsquerda=$v->[$meio];

  #para todas as posicoes a esquerda
  for(my $i=$meio-1;$i>=$esquerda;--$i)
  {
    $s+=$v->[$i];
    if($s>$rEsquerda)
    {
      $rEsquerda=$s;
    }
  }

  #inicia soma parcial direita
  $s=$v->[$meio+1];

  #inicial resultado parcial direito
  my $rDireita=$v->[$meio+1];

  #para todas as posicoes a direita
  for(my $i=$meio+2;$i<=$direita;++$i)
  {
    $s+=$v->[$i];
    if($s>$rDireita)
    {
      $rDireita=$s;
    }
  }

  my $r=&maximo($somaMaximaEsquerda,$rEsquerda+$rDireita,$somaMaximaDireita);
  return $r;
}

sub maximo
{
  if(!@_)
  {
    return undef;
  }
  my $max=shift;
  while(@_)
  {
    my $x=shift;
    if($x>$max)
    {
      $max=$x;
    }
  }
  return $max;
}

sub parseARGV
{
  my $arg=shift;

  if(!defined $arg)
  {
    return -1;
  }

  if($arg=~/^-?\d+$/)
  {
    return 1;
  }

  my $infile;
  if($arg=~/-?-?[Ss][Tt][Dd](?:[Ii][Nn])?$/)
  {
    $infile=*STDIN;
  }
  else
  {
    if(!open($infile,'<',$arg))
    {
      return -2;
    }
  }
  my $c=$ARGV[1] // 0;
  @ARGV=();
  if((defined $c)&&($c=~/^[1-9]\d*$/))
  {
    $#ARGV=$c-1;
  }
  my $x=0;
  while(my $line=<$infile>)
  {
    chomp($line);
    #push @ARGV, $line;
    $ARGV[$x++]=int($line);
  }
  if($x>0)
  {
    $#ARGV=$x-1;
  }
  return 2;
}
